File: src/rust/src/z_algorithm.rs
pub fn z_algorithm<T>(a: &[T]) -> Vec<usize>
where
    T: PartialEq,
{
    let n = a.len();
    let mut lcp = vec![0; n];
    let mut l = 0;
    for i in 1..n {
        let r = l + lcp[l];
        let mut d = if r <= i { 0 } else { lcp[i - l].min(r - i) };
        while i + d < n && a[i + d] == a[d] {
            d += 1;
        }
        lcp[i] = d;
        if r < i + d {
            l = i;
        }
    }
    lcp[0] = n;
    lcp
}

File: src/python/src/dsalgo/z_algorithm.py
import typing

T = typing.TypeVar("T")


def z_algorithm(a: typing.Sequence[T]) -> typing.List[int]:
    n = len(a)
    lcp = [0] * n
    l = 0
    for i in range(1, n):
        r = l + lcp[l]
        d = 0 if r <= i else min(lcp[i - l], r - i)
        while i + d < n and a[i + d] == a[d]:
            d += 1
        lcp[i] = d
        if i + d > r:
            l = i
    lcp[0] = n
    return lcp
